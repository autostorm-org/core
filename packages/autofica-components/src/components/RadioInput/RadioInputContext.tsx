import React from "react";
interface IRadioContext {
  value: string;
  setValue: (_value: string) => void;
  name: string;
}

const radioContextInitialValue: IRadioContext = {
  value: "",
  setValue: (_value: string) => {},
  name: "Radio group",
};

const RadioContext = React.createContext<IRadioContext>(
  radioContextInitialValue
);

interface IRadioGroupProviderProps {
  name: string;
  initialValue: string;
  onValueChange?: (nextValue: string) => void;
}

const useValueChangeHandler = (
  setValue: React.Dispatch<React.SetStateAction<string>>,
  onChange?: (nextValue: string) => void
) => {
  return React.useCallback(
    (value: string) => {
      setValue(value);
      if (onChange != null) {
        onChange(value);
      }
    },
    [setValue, onChange]
  );
};

const RadioGroupProvider = (
  props: React.PropsWithChildren<IRadioGroupProviderProps>
) => {
  const [value, setValue] = React.useState(props.initialValue);
  const onValueChange = useValueChangeHandler(setValue, props.onValueChange);
  return (
    <RadioContext.Provider
      value={{ value, setValue: onValueChange, name: props.name }}
    >
      {props.children}
    </RadioContext.Provider>
  );
};

export default RadioGroupProvider;
export { RadioContext, radioContextInitialValue };
export type { IRadioGroupProviderProps, IRadioContext };
