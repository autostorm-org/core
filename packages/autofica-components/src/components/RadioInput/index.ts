//RadioInput
export { default as RadioInput } from "./RadioInput";
export type { IRadioInputProps } from "./RadioInput";
// RadioGroupProvider
export { default as RadioGroupProvider } from "./RadioInputContext";
export { RadioContext, radioContextInitialValue } from "./RadioInputContext";
export type {
  IRadioGroupProviderProps,
  IRadioContext,
} from "./RadioInputContext";
