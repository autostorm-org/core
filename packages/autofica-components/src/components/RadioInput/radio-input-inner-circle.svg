<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 778.51 778.51"><g id="Layer_2" data-name="Layer 2"><g id="Layer_1-2" data-name="Layer 1"><path d="M389.25,0C174.39,0,0,174.39,0,389.25S174.39,778.51,389.25,778.51,778.51,604.12,778.51,389.25,604.12,0,389.25,0Z"/></g></g></svg>