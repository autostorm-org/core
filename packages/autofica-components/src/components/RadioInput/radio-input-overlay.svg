<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1557.02 1557.02"><defs><style>.cls-1{fill:none;}</style></defs><g id="Layer_2" data-name="Layer 2"><g id="Layer_1-2" data-name="Layer 1"><path class="cls-1" d="M778.51,0C348.38,0,0,348.38,0,778.51S348.38,1557,778.51,1557,1557,1208.63,1557,778.51,1208.63,0,778.51,0Zm0,1401.31c-344.1,0-622.81-278.7-622.81-622.8S434.41,155.7,778.51,155.7s622.8,278.71,622.8,622.81S1122.61,1401.31,778.51,1401.31Z"/><path d="M778.08,389.25c-214.86,0-389.25,174.39-389.25,389.26s174.39,389.25,389.25,389.25,389.26-174.38,389.26-389.25S993,389.25,778.08,389.25Z"/></g></g></svg>