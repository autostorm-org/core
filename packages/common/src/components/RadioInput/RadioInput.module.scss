.root {
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  transition: 0.2s;
  transition-property: box-shadow;
  cursor: pointer;
  &:focus,
  &:focus-within {
    box-shadow: 0 0 0 0.2rem var(--button-box-shadow-color, #b3b3b3);
    outline: none;
  }

  & input:not(:checked) ~ svg {
    filter: var(--radio-input-fitler-unchecked);
  }
  & input:checked ~ svg {
    filter: var(--radio-input-filter-checked);
  }

  & input:disabled ~ svg {
    filter: var(--radio-input-fitler-disabled);
    cursor: none;
  }

  & input:checked:disabled ~ svg {
    filter: var(--radio-input-filter-checked-disabled);
  }
}

.disabledText {
  opacity: 0.8;
}

.input {
  top: 0;
  left: 0;
  width: 100%;
  cursor: inherit;
  height: 100%;
  margin: 0;
  opacity: 0;
  padding: 0;
  z-index: 1;
  position: absolute;
}

.outerCircle {
  width: 20px;
}

.overlay {
  width: 20px;
  position: absolute;
  left: 0;
  transform: scale(0);
  transition: 0.2s;
  transition-property: transform;
}
.overlayVisible {
  transform: scale(1);
}

.svgContainer {
  margin-right: 5px;
  display: inline-block;
  vertical-align: middle;
  position: relative;
}
