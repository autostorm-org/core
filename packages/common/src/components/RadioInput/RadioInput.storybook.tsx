import React from "react";
import RadioInput from "./RadioInput";
import RadioInputContextProvider from "./RadioInputContext";
import { Card } from "../Card";
import { Typography, TypographyVariant_ENUM } from "../Typography";
export default {
  title: "RadioInput Story",
  component: RadioInput,
};

export const RadioInputStory = () => {
  return (
    <Card>
      <div role="radiogroup" aria-labelledby="Greeting Title">
        <Typography variant={TypographyVariant_ENUM.h3} id="Greeting Title">
          Chose your greeting
        </Typography>

        <br></br>
        <RadioInputContextProvider name={"Greetings"} initialValue={"hello"}>
          <RadioInput
            name={"greeting"}
            value={"hello"}
            id={"Hello Option"}
            disabled={true}
            rightVisualLabel={
              <Typography variant={TypographyVariant_ENUM.p}>hello</Typography>
            }
          />
          <br></br>

          <RadioInput
            name={"greeting"}
            value={"world"}
            id={"world Option"}
            rightVisualLabel={
              <Typography variant={TypographyVariant_ENUM.p}>world</Typography>
            }
          />
          <br></br>
        </RadioInputContextProvider>
      </div>

      <div role="radiogroup" aria-label="Chose your pet">
        <Typography variant={TypographyVariant_ENUM.h3}>
          <label>Chose your Pet</label>
        </Typography>

        <br></br>
        <RadioInputContextProvider name={"Pet"} initialValue={""}>
          <RadioInput
            name={"Pet"}
            value={"Akita Inu"}
            id={"Akita Inu Option"}
            rightVisualLabel={
              <Typography variant={TypographyVariant_ENUM.p}>
                Akita Inu
              </Typography>
            }
          />
          <br></br>

          <RadioInput
            name={"Pet"}
            value={"Shiba Inu"}
            id={"Shiba Inu Option"}
            rightVisualLabel={
              <Typography variant={TypographyVariant_ENUM.p}>
                Shiba Inu
              </Typography>
            }
          />
          <br></br>
        </RadioInputContextProvider>
      </div>
    </Card>
  );
};
