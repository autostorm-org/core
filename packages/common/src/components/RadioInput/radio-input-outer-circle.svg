<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1557.02 1557.02"><g id="Layer_2" data-name="Layer 2">
<g id="Layer_1-2" data-name="Layer 1">
<path d="M778.51,0C348.38,0,0,348.38,0,778.51S348.38,1557,778.51,1557,1557,1208.63,1557,778.51,1208.63,0,778.51,0Zm0,1401.31c-344.1,0-622.81-278.7-622.81-622.8S434.41,155.7,778.51,155.7s622.8,278.71,622.8,622.81S1122.61,1401.31,778.51,1401.31Z"/>
</g>
</g>
</svg>