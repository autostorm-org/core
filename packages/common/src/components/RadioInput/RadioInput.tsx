import React from "react";
import { ReactComponent as OuterCircle } from "./radio-input-outer-circle.svg";
import { ReactComponent as Overlay } from "./radio-input-overlay.svg";
import { RadioContext } from "./RadioInputContext";
import type {
  IIdentifiable,
  IAriaLabel,
  IChangeable,
  IFocusable,
  IStylable,
  IOverridable,
} from "../../types";
import styles from "./RadioInput.module.scss";

const RADIOTYPE = "radio" as "radio";

interface IRadioInputProps
  extends IAriaLabel,
    IChangeable<HTMLInputElement>,
    IFocusable<HTMLInputElement>,
    IIdentifiable,
    IStylable,
    IOverridable {
  name: string;
  value: string;
  checked?: boolean;
  leftVisualLabel?: React.ReactNode;
  rightVisualLabel?: React.ReactNode;
  disabled?: boolean;
}

const useOnChangeHandler = (setValue: (_value: string) => void) => {
  return React.useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      setValue(e.target.value);
    },
    [setValue]
  );
};

const RadioInput = React.forwardRef<HTMLLabelElement, IRadioInputProps>(
  (props, ref) => {
    const radioContext = React.useContext(RadioContext);
    const onChange = useOnChangeHandler(radioContext.setValue);
    const overlayClassName = `${styles.overlay} ${
      radioContext.value == props.value ? styles.overlayVisible : ""
    }`;
    return (
      <label
        role={RADIOTYPE}
        className={`${styles.root} ${
          props.disabled ? styles.disabledText : ""
        }`}
        htmlFor={props.id}
        aria-label={props["aria-label"]}
        aria-checked={radioContext.value == props.value}
        ref={ref}
      >
        {props.leftVisualLabel}
        <span className={styles.svgContainer}>
          <input
            className={styles.input}
            type={RADIOTYPE}
            id={props.id}
            onChange={onChange}
            onBlur={props.onBlur}
            onFocus={props.onFocus}
            onFocusCapture={props.onFocusCapture}
            name={radioContext.name}
            value={props.value}
            checked={radioContext.value == props.value}
            disabled={props.disabled}
          ></input>

          <OuterCircle className={styles.outerCircle} />
          <Overlay className={overlayClassName} />
        </span>
        {props.rightVisualLabel}
      </label>
    );
  }
);
export default RadioInput;
export { RadioContext };
export type { IRadioInputProps };
