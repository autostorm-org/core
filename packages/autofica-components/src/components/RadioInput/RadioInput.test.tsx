import React from "react";
import renderer from "react-test-renderer";
import type { ReactTestRenderer } from "react-test-renderer";
import RadioInput from "./RadioInput";
import RadioInputContextProvider from "./RadioInputContext";

describe("Radio Button group Tests", () => {
  let root: ReactTestRenderer = null;
  afterEach(() => {
    if (root != null) {
      root.unmount();
    }
  });

  it("renders radio inputs", () => {
    root = renderer.create(<RadioInput name="group" value="test"></RadioInput>);
    expect(root.toJSON()).toMatchSnapshot();
  });

  it("renders renders a radio group", () => {
    root = renderer.create(
      <RadioInputContextProvider name="group" initialValue="">
        <RadioInput name="group" value="a"></RadioInput>
        <RadioInput name="group" value="b"></RadioInput>
      </RadioInputContextProvider>
    );
    expect(root.toJSON()).toMatchSnapshot();
  });
});
