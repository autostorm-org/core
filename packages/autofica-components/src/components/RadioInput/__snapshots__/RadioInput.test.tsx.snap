// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio Button group Tests renders radio inputs 1`] = `
<label
  aria-checked={false}
  className="root "
  role="radio"
>
  <span
    className="svgContainer"
  >
    <input
      checked={false}
      className="input"
      name="Radio group"
      onChange={[Function]}
      type="radio"
      value="test"
    />
    <svg-mock
      className="outerCircle"
    />
    <svg-mock
      className="overlay "
    />
  </span>
</label>
`;

exports[`Radio Button group Tests renders renders a radio group 1`] = `
Array [
  <label
    aria-checked={false}
    className="root "
    role="radio"
  >
    <span
      className="svgContainer"
    >
      <input
        checked={false}
        className="input"
        name="group"
        onChange={[Function]}
        type="radio"
        value="a"
      />
      <svg-mock
        className="outerCircle"
      />
      <svg-mock
        className="overlay "
      />
    </span>
  </label>,
  <label
    aria-checked={false}
    className="root "
    role="radio"
  >
    <span
      className="svgContainer"
    >
      <input
        checked={false}
        className="input"
        name="group"
        onChange={[Function]}
        type="radio"
        value="b"
      />
      <svg-mock
        className="outerCircle"
      />
      <svg-mock
        className="overlay "
      />
    </span>
  </label>,
]
`;
